stdlib/REPL/src/REPL.jl

```diff
@@ -116,6 +117,9 @@ import Base:
     AnyDict,
     ==
 
+show(io::IO, ::MIME"text/plain", ex::Expr) =
+    print(io, JuliaSyntaxHighlighting.highlight(sprint(show, ex)))
```

Comment:
This is type-piracy so I guess this should be caught higher up in the REPLDisplay code somewhere?

Reply:
I suppose line 120 doesn't look that high up, but it is just after the `using` and `exports` lines.

Reply:
Type piracy means that one defines a method where one neither owns the function or any of the types (which is the case here).

With "higher up" it doesn't mean higher up in the file but higher up in the "dependency chain" (here this means in Base since Base owns `show` and `Expr`).

Reply:
The REPL stdlib has it's own display and I guess this is something that should be implemented there, i.e. how REPL `display` `Exprs` and not how they are `show`d.

Reply:
Yea, doing this in `Base` isn't really an option since we're using two stdlibs here.

Reply:
Type piracy like this is also not an option though.

As @fredrikekre suggested, an alternative is to add something to https://github.com/JuliaLang/julia/blob/8742d3ccd9fb9a10d4147eada774ba923c3faac3/stdlib/REPL/src/REPL.jl#L383 to special case `Expr`.

Reply:
I'll just switch from implementing a `show` method to a `display` method then :)

Reply:
Hehe, `display` is also a Base function. Instead of

```julia
 show(io, mime, x[]) 
```

in the REPL you could add something like 

```julia
...
if x[] is a Expr
   print(io, JuliaSyntaxHighlighting.highlight(sprint(show, ex)))
else
  show(io, mime, x[]) 
end
...
```

to https://github.com/JuliaLang/julia/blob/8742d3ccd9fb9a10d4147eada774ba923c3faac3/stdlib/REPL/src/REPL.jl#L383

Reply:
The display type is owned by REPL though, but I think perhaps in this case it is cleaner to change https://github.com/JuliaLang/julia/blob/8742d3ccd9fb9a10d4147eada774ba923c3faac3/stdlib/REPL/src/REPL.jl#L383 to 
```julia
show_repl(io, mime, x[])
```
with implementations
```julia
show_repl(io::IO, mime::MIME, x) = show(io, mime, x)
show_repl(io::IO, mime, x::Expr) = ...
```
?

Reply:
> The display type is owned by REPL though

Yes, it's the function I am suggesting to edit here heh, but it sets up various iocontexts etc which is also why adding a new overload to it isn't that good.

It would also be good to check that https://github.com/JuliaLang/julia/pull/38028 doesn't regress.

Reply:
I've just pushed the change that @fredrikekre suggested :slightly_smiling_face: